infra/src/main/kotlin/fi/fta/geoviite/api/tracklayout/v1/ExtTrackNumberCollectionServiceV1.kt

```diff
@@ -0,0 +1,184 @@
+    private val layoutTrackNumberDao: LayoutTrackNumberDao,
+    private val publicationDao: PublicationDao,
+    private val publicationService: PublicationService,
+    private val referenceLineService: ReferenceLineService,
+) {
+    val logger: Logger = LoggerFactory.getLogger(this::class.java)
+
+    fun createTrackNumberCollectionResponse(
+        trackLayoutVersion: Uuid<Publication>?,
+        coordinateSystem: Srid,
+    ): ExtTrackNumberCollectionResponseV1 {
+        val layoutContext = MainLayoutContext.official
+
+        val (publication, locationTracks) =
+            when (trackLayoutVersion) {
+                null -> {
+                    publicationDao.fetchLatestPublications(LayoutBranchType.MAIN, count = 1).single().let {
+                        newestPublication ->
+                        newestPublication to layoutTrackNumberDao.list(layoutContext, includeDeleted = false)
+                    }
+                }
+                else -> {
+                    trackLayoutVersion
+                        .let { uuid -> publicationDao.fetchPublicationByUuid(uuid) }
+                        ?.let { specifiedPublication ->
+                            specifiedPublication to
+                                layoutTrackNumberDao.listPublishedLayoutTrackNumbersAtMoment(
+                                    specifiedPublication.publicationTime
+                                )
+                        } ?: throw TrackLayoutVersionNotFound("trackLayoutVersion=${trackLayoutVersion}")
+                }
+            }
+
+        return ExtTrackNumberCollectionResponseV1(
+            trackLayoutVersion = publication.uuid,
+            coordinateSystem = coordinateSystem,
+            trackNumberCollection =
+                extGetTrackNumberCollection(
+                    layoutContext,
+                    locationTracks,
+                    coordinateSystem,
+                    publication.publicationTime,
+                ),
+        )
+    }
+
+    fun createTrackNumberCollectionModificationResponse(
+        modificationsFromVersion: Uuid<Publication>,
+        trackLayoutVersion: Uuid<Publication>?,
+        coordinateSystem: Srid,
+    ): ExtModifiedTrackNumberCollectionResponseV1? {
+        val layoutContext = MainLayoutContext.official
+
+        val (fromPublication, toPublication) =
+            publicationService.getPublicationsToCompare(modificationsFromVersion, trackLayoutVersion)
+
+        return if (fromPublication == toPublication) {
+            logger.info(
+                "there cannot be any differences if the requested publication ids are the same, publicationId=${fromPublication.id}"
+            )
+            null
+        } else {
+            val changedIds =
+                publicationDao.fetchPublishedTrackNumbersAfterMoment(
+                    fromPublication.publicationTime,
+                    toPublication.publicationTime,
+                )
+
+            val modifiedTrackNumbers =
```

Comment:
Yllä haetaan nyt ensin varta vasten pelkät ID:t, ignoraten versiot ja sitten haetaan ajanhetkille versiot->oliot. Tuossa ylemmässä on toki se että se hakee kaikki ID:t jotka on ko. välillä jossain kohtaa muuttunut (johonkin voi olla useampi muutos), mutta kun se kerran on tätä tiettyä asiaa varten oleva täsmä-haku niin voisi varmaan olla paremmin tarkoitukseen sopiva. Esim jos se palauttaisi vain from&to versiot suoraan kullekin noista muuttuneista (versiotiedothan on jo julkaisuriveissä mukana) niin tässä voisi sitten vain hakea versioita vastaavat oliot kerralla daolta fetchMany:llä.

Reply:
Eikunjaa, vielä se että eihän tuo edes tee diffiä eikä siksi toki hae niitä vanhoja lainkaan. Vielä helpompaa -> riittää ottaa ne uudet versiot.